Save manager improvements

MSuih opened the pull request:
* Display icons.
* Show details in the main UI instead of popup dialog.
* Various smaller tweaks to interface and codebase.

![image](https://user-images.githubusercontent.com/19759300/61928528-63b59f80-af81-11e9-8c5d-c60e25d463b7.png)

I had to combine title and subtitle into one field because the interface would've been too wide with them as separate fields, hopefully this doesn't upset anyone.

Megamouse commented on rpcs3/rpcs3qt/save_manager_dialog.h:
```diff
@@ -40,4 +42,15 @@ private Q_SLOTS:
 	bool m_sort_ascending;
+	QSize m_icon_size;
+
+	QLabel* m_details_icon;
```
= nullptr

Megamouse commented on rpcs3/rpcs3qt/save_manager_dialog.cpp:
```diff
@@ -273,66 +344,18 @@ void save_manager_dialog::OnEntriesRemove()
-void save_manager_dialog::ShowContextMenu(const QPoint &pos)
+void save_manager_dialog::SetIconSize(int size)
 {
-	bool selectedItems = m_list->selectionModel()->selectedRows().size() > 1;
```
Why did you remove the contextmenu?
People are gonna add it back.

MSuih replied:
I don't like having same features in two places, everything in the context menu is accessible from other places in the interface. But I can add it back if that is desired.

Megamouse replied:
Most will say yes

Megamouse commented on rpcs3/rpcs3qt/save_manager_dialog.cpp:
```diff
@@ -160,35 +219,49 @@ void save_manager_dialog::UpdateList()
+
 		QString title = qstr(entry.title);
-		QString subtitle = qstr(entry.subtitle);
+		title.append("\n");
```
Can't this be done in the initialization?

Megamouse commented on rpcs3/rpcs3qt/save_manager_dialog.cpp:
```diff
@@ -160,35 +219,49 @@ void save_manager_dialog::UpdateList()
-		QTableWidgetItem* subtitleItem = new QTableWidgetItem(subtitle);
-		subtitleItem->setFlags(subtitleItem->flags() & ~Qt::ItemIsEditable);
-		m_list->setItem(row, 1, subtitleItem);
+		QTableWidgetItem* modifiedItem = new QTableWidgetItem(FormatTimestamp(entry.mtime));
```
Just my personal preference, but wouldn't it be better to call this timeItem or something like that?

Megamouse commented on rpcs3/rpcs3qt/save_manager_dialog.cpp:
```diff
@@ -203,6 +276,19 @@ void save_manager_dialog::UpdateList()
+	for (int i = 0; i < m_list->rowCount(); i++)
+	{
+		QTableWidgetItem* item = m_list->item(i, 0);
+		QPixmap data = item->data(Qt::UserRole).value<QPixmap>();
```
I was wondering if these won't get too pixelated in hidpi modes

MSuih replied:
I don't have access to hidpi screen so I can't test it, but theoretically this should be fine.

Megamouse replied:
I did this in the game list and it was much appreciated

Megamouse commented on rpcs3/rpcs3qt/save_manager_dialog.cpp:
```diff
@@ -100,47 +107,99 @@ void save_manager_dialog::Init(std::string dir)
-	// Button Layout
+	// Details
+	m_details_icon = new QLabel(this);
+	m_details_icon->setMinimumSize(320, 176);
```
Just wondering why you didn't just use a new form file.
Most of these settings can be done with simple mouse clicks

MSuih replied:
I just started with small tweaks to the current UI and it grew from there line by line. Besides I'm used to creating interfaces with code rather than external editor.

Megamouse replied:
That's fine. As I said, I was just curious.

Megamouse commented:
Also yeah... The merging of two columns is probably going to upset at least one person.

MSuih commented:
I don't see a way of remedying that without cutting out something else from the interface. I tried having them as separate columns and the interface was almost wide enough to cover a 1080p monitor from side to side. The subtitles are rarely useful as standalone items anyways, most of them tend to be some variation of "Save Data".

vsub commented:
Looks nice
Can you also make the save\load dialogs looks the same way(see an icon and not having to double click to see the save info)while using QT,not the native interface

Megamouse commented:
@vsub those are just a fallback but if he wants to do it then that's cool too.

Megamouse commented:
@MSuih let's just say the subs are artistic freedom :)

MSuih commented:
Added back most of the rightclick menu and added pixel ratio stuff to image rendering. Hopefully the DPI stuff works, I tested it by scaling one of my monitors.

@vsub Not interested right now and I don't want to make any promises about future either.

Megamouse commented on rpcs3/rpcs3qt/save_manager_dialog.cpp:
```diff
@@ -203,6 +278,27 @@ void save_manager_dialog::UpdateList()
 	resize(preferredSize.boundedTo(maxSize));
 }
 
+void save_manager_dialog::UpdateIcons()
```
```suggestion
void save_manager_dialog::UpdateIcons()
{
	const int dpr = devicePixelRatio();
	const int width = m_icon_size.width() * dpr;
	const int height = m_icon_size.height() * dpr * 176 / 320;

	for (int i = 0; i < m_list->rowCount(); i++)
	{
		QTableWidgetItem* item = m_list->item(i, 0);
		QPixmap data = item->data(Qt::UserRole).value<QPixmap>();

		QPixmap icon(data.size() * dpr);
		icon.setDevicePixelRatio(dpr);
		icon.fill(Qt::transparent);

		QPainter painter(&icon);
		painter.drawPixmap(0, 0, data);

		item->setData(Qt::DecorationRole, icon.scaled(width, height, Qt::KeepAspectRatio, Qt::TransformationMode::SmoothTransformation));
	}

	m_list->resizeRowsToContents();
	m_list->resizeColumnToContents(0);
}
```

Megamouse commented:
Just for comparison:

![Comparison](https://user-images.githubusercontent.com/23019877/61991648-8912de00-b053-11e9-82b9-14a914146319.png)

MSuih commented:
Thanks, I tried to do something like that but I didn't realize I have to first draw it to high-dpi surface and then scale the end result back. Qt's implementation of highdpi support is confusing.

Megamouse commented on rpcs3/rpcs3qt/save_manager_dialog.cpp:
```diff
@@ -203,6 +278,28 @@ void save_manager_dialog::UpdateList()
+{
+	const int dpr = devicePixelRatio();
+	const int width = m_icon_size.width() * dpr;
+	const int height = m_icon_size.height() *dpr * 176 / 320;
```
I don't wanna be an ass but this is missing a space :(

MSuih replied:
Sigh.

Megamouse commented:
Yeah. This isn't a problem if you use resources and forms.
But since we are doing this dynamically we have to set this up correctly.
Took me a while to get it right...
